File: frontend/src/utils/cnpj.js
/**
 * Classe utilitária para validação e formatação de CNPJ no frontend.
 * Espelha a lógica do backend para dar feedback imediato ao usuário.
 */
export class CnpjUtil {
  static _limpar(cnpj) {
    return cnpj.replace(/\D/g, '')
  }

  static _calcularDigito(cnpjParcial, pesos) {
    const soma = cnpjParcial
      .split('')
      .reduce((acc, d, i) => acc + parseInt(d) * pesos[i], 0)
    const resto = soma % 11
    return resto < 2 ? 0 : 11 - resto
  }

  static validar(cnpj) {
    cnpj = this._limpar(cnpj)
    if (cnpj.length !== 14) return false
    if (/^(\d)\1+$/.test(cnpj)) return false

    const pesos1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    const pesos2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]

    const d1 = this._calcularDigito(cnpj.slice(0, 12), pesos1)
    const d2 = this._calcularDigito(cnpj.slice(0, 13), pesos2)

    return cnpj.slice(-2) === `${d1}${d2}`
  }

  static formatar(cnpj) {
    cnpj = this._limpar(cnpj)
    if (cnpj.length !== 14) return cnpj
    return `${cnpj.slice(0, 2)}.${cnpj.slice(2, 5)}.${cnpj.slice(5, 8)}/${cnpj.slice(8, 12)}-${cnpj.slice(12)}`
  }

  /** Aplica máscara enquanto o usuário digita */
  static mascarar(cnpj) {
    cnpj = this._limpar(cnpj).slice(0, 14)
    if (cnpj.length <= 2) return cnpj
    if (cnpj.length <= 5) return `${cnpj.slice(0, 2)}.${cnpj.slice(2)}`
    if (cnpj.length <= 8) return `${cnpj.slice(0, 2)}.${cnpj.slice(2, 5)}.${cnpj.slice(5)}`
    if (cnpj.length <= 12) return `${cnpj.slice(0, 2)}.${cnpj.slice(2, 5)}.${cnpj.slice(5, 8)}/${cnpj.slice(8)}`
    return `${cnpj.slice(0, 2)}.${cnpj.slice(2, 5)}.${cnpj.slice(5, 8)}/${cnpj.slice(8, 12)}-${cnpj.slice(12)}`
  }
}

File: backend/app/utils/cnpj.py
"""
Utilitário de validação e formatação de CNPJ usando Programação Orientada a Objetos.
"""
import re


class CnpjValidator:
    """
    Classe responsável por validar e formatar CNPJs.
    Aplica o algoritmo oficial dos dois dígitos verificadores.

    Uso:
        CnpjValidator.validar("11222333000181")  -> True / False
        CnpjValidator.formatar("11222333000181") -> "11.222.333/0001-81"
    """

    _PESOS_PRIMEIRO_DIGITO = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    _PESOS_SEGUNDO_DIGITO = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]

    @classmethod
    def _limpar(cls, cnpj: str) -> str:
        """Remove todos os caracteres não numéricos."""
        return re.sub(r"\D", "", cnpj)

    @classmethod
    def _calcular_digito(cls, cnpj_parcial: str, pesos: list[int]) -> int:
        """Calcula um dígito verificador com base nos pesos fornecidos."""
        soma = sum(int(d) * p for d, p in zip(cnpj_parcial, pesos))
        resto = soma % 11
        return 0 if resto < 2 else 11 - resto

    @classmethod
    def validar(cls, cnpj: str) -> bool:
        """
        Valida um CNPJ (com ou sem formatação).
        Retorna True se o CNPJ for matematicamente válido.
        """
        cnpj = cls._limpar(cnpj)

        if len(cnpj) != 14:
            return False

        # Rejeita sequências de dígitos todos iguais (ex: 00000000000000)
        if len(set(cnpj)) == 1:
            return False

        primeiro = cls._calcular_digito(cnpj[:12], cls._PESOS_PRIMEIRO_DIGITO)
        segundo = cls._calcular_digito(cnpj[:13], cls._PESOS_SEGUNDO_DIGITO)

        return cnpj[-2:] == f"{primeiro}{segundo}"

    @classmethod
    def formatar(cls, cnpj: str) -> str:
        """
        Formata um CNPJ para o padrão XX.XXX.XXX/XXXX-XX.
        Lança ValueError se o CNPJ não tiver 14 dígitos após limpeza.
        """
        cnpj = cls._limpar(cnpj)
        if len(cnpj) != 14:
            raise ValueError("CNPJ deve conter exatamente 14 dígitos.")
        return f"{cnpj[:2]}.{cnpj[2:5]}.{cnpj[5:8]}/{cnpj[8:12]}-{cnpj[12:]}"
